File: libs/python/tests/test_output.py
import os

import yaml
import pytest

from spade._output import _infer_output_name, read_block_manifest, write_outputs
from spade.types import (
    Directory,
    File,
    FileCollection,
    JsonFile,
    RasterFile,
    RasterFileCollection,
    VectorFile,
)


class TestInferOutputName:
    def test_file(self):
        assert _infer_output_name(File(path="/tmp/a")) == "file"

    def test_raster_file(self):
        assert _infer_output_name(RasterFile(path="/tmp/a")) == "raster"

    def test_vector_file(self):
        assert _infer_output_name(VectorFile(path="/tmp/a")) == "vector"

    def test_json_file(self):
        assert _infer_output_name(JsonFile(path="/tmp/a")) == "json"

    def test_file_collection(self):
        assert _infer_output_name(FileCollection(paths=[])) == "files"

    def test_raster_file_collection(self):
        assert _infer_output_name(RasterFileCollection(paths=[])) == "rasters"


class TestWriteOutputs:
    def test_none_result(self, work_dir):
        write_outputs(None)
        assert list((work_dir / "outputs").iterdir()) == []

    def test_single_file_output(self, work_dir):
        src = work_dir / "tmp_result.tif"
        src.write_bytes(b"raster data")

        write_outputs(RasterFile(path=str(src)))
        output_dir = work_dir / "outputs" / "raster"
        assert output_dir.exists()
        assert (output_dir / "tmp_result.tif").exists()
        assert (output_dir / "tmp_result.tif").read_bytes() == b"raster data"

    def test_single_file_with_manifest(self, work_dir):
        src = work_dir / "result.tif"
        src.write_bytes(b"data")
        manifest = {"custom_output": {"type": "file", "format": "GeoTIFF"}}
        write_outputs(RasterFile(path=str(src)), manifest_outputs=manifest)
        assert (work_dir / "outputs" / "custom_output" / "result.tif").exists()

    def test_dict_output(self, work_dir):
        src_raster = work_dir / "result.tif"
        src_raster.write_bytes(b"raster")
        src_json = work_dir / "summary.json"
        src_json.write_bytes(b'{"key": "value"}')

        write_outputs({
            "raster": RasterFile(path=str(src_raster)),
            "summary": JsonFile(path=str(src_json)),
        })

        assert (work_dir / "outputs" / "raster" / "result.tif").exists()
        assert (work_dir / "outputs" / "summary" / "summary.json").exists()

    def test_collection_output(self, work_dir):
        paths = []
        for i in range(3):
            p = work_dir / f"tile_{i}.tif"
            p.write_bytes(f"tile {i}".encode())
            paths.append(str(p))

        write_outputs(RasterFileCollection(paths=paths))
        output_dir = work_dir / "outputs" / "rasters"
        assert output_dir.exists()
        assert len(list(output_dir.iterdir())) == 3

    def test_directory_output(self, work_dir):
        src_dir = work_dir / "result_dir"
        src_dir.mkdir()
        (src_dir / "file1.txt").write_bytes(b"a")
        (src_dir / "file2.txt").write_bytes(b"b")

        write_outputs(Directory(path=str(src_dir)))
        output_dir = work_dir / "outputs" / "directory"
        assert output_dir.exists()
        assert (output_dir / "file1.txt").exists()
        assert (output_dir / "file2.txt").exists()

    def test_preserves_filename(self, work_dir):
        src = work_dir / "my_custom_name.geojson"
        src.write_bytes(b"geojson data")
        write_outputs(VectorFile(path=str(src)))
        assert (work_dir / "outputs" / "vector" / "my_custom_name.geojson").exists()


class TestReadBlockManifest:
    def test_no_manifest(self, work_dir):
        assert read_block_manifest() is None

    def test_block_yaml_in_cwd(self, work_dir):
        manifest = {
            "id": "test.block",
            "outputs": {"raster": {"type": "file", "format": "GeoTIFF"}},
        }
        with open(work_dir / "block.yaml", "w") as f:
            yaml.dump(manifest, f)
        result = read_block_manifest()
        assert result == {"raster": {"type": "file", "format": "GeoTIFF"}}

    def test_env_var_manifest(self, work_dir, tmp_path):
        manifest_path = tmp_path / "external" / "block.yaml"
        manifest_path.parent.mkdir(exist_ok=True)
        manifest = {
            "outputs": {"output": {"type": "file"}},
        }
        with open(manifest_path, "w") as f:
            yaml.dump(manifest, f)
        os.environ["SPADE_BLOCK_MANIFEST"] = str(manifest_path)
        try:
            result = read_block_manifest()
            assert result == {"output": {"type": "file"}}
        finally:
            del os.environ["SPADE_BLOCK_MANIFEST"]

File: libs/R/tests/testthat/test-output.R
test_that("infer_output_name for File", {
  expect_equal(infer_output_name(File(path = "/tmp/x")), "file")
})

test_that("infer_output_name for RasterFile", {
  expect_equal(infer_output_name(RasterFile(path = "/tmp/x")), "raster")
})

test_that("infer_output_name for VectorFile", {
  expect_equal(infer_output_name(VectorFile(path = "/tmp/x")), "vector")
})

test_that("infer_output_name for JsonFile", {
  expect_equal(infer_output_name(JsonFile(path = "/tmp/x")), "json")
})

test_that("infer_output_name for FileCollection", {
  expect_equal(infer_output_name(FileCollection(paths = "/tmp/x")), "files")
})

test_that("infer_output_name for RasterFileCollection", {
  expect_equal(infer_output_name(RasterFileCollection(paths = "/tmp/x")), "rasters")
})

test_that("write_outputs NULL result produces no output", {
  env <- setup_work_dir()
  on.exit(teardown_work_dir(env))

  write_outputs(NULL)
  # outputs dir should not exist (or be empty if pre-created)
  output_subdirs <- list.dirs("outputs", recursive = FALSE)
  expect_equal(length(output_subdirs), 0)
})

test_that("write_outputs single RasterFile writes to outputs/raster/", {
  env <- setup_work_dir()
  on.exit(teardown_work_dir(env))

  # Create a source file
  src <- tempfile(fileext = ".tif")
  writeBin(charToRaw("raster data"), src)

  result <- RasterFile(path = src)
  write_outputs(result)

  expect_true(dir.exists("outputs/raster"))
  output_files <- list.files("outputs/raster")
  expect_equal(length(output_files), 1)
})

test_that("write_outputs single file with manifest uses manifest name", {
  env <- setup_work_dir()
  on.exit(teardown_work_dir(env))

  src <- tempfile(fileext = ".tif")
  writeBin(charToRaw("raster data"), src)

  result <- RasterFile(path = src)
  manifest_outputs <- list(custom_output = list(type = "file"))
  write_outputs(result, manifest_outputs)

  expect_true(dir.exists("outputs/custom_output"))
})

test_that("write_outputs named list writes multiple subdirectories", {
  env <- setup_work_dir()
  on.exit(teardown_work_dir(env))

  src1 <- tempfile(fileext = ".tif")
  src2 <- tempfile(fileext = ".geojson")
  writeBin(charToRaw("raster"), src1)
  writeBin(charToRaw("vector"), src2)

  result <- list(
    raster_out = RasterFile(path = src1),
    vector_out = VectorFile(path = src2)
  )
  write_outputs(result)

  expect_true(dir.exists("outputs/raster_out"))
  expect_true(dir.exists("outputs/vector_out"))
})

test_that("write_outputs collection writes all files", {
  env <- setup_work_dir()
  on.exit(teardown_work_dir(env))

  src1 <- tempfile(tmpdir = tempdir(), fileext = ".tif")
  src2 <- tempfile(tmpdir = tempdir(), fileext = ".tif")
  writeBin(charToRaw("raster1"), src1)
  writeBin(charToRaw("raster2"), src2)

  result <- RasterFileCollection(paths = c(src1, src2))
  write_outputs(result)

  expect_true(dir.exists("outputs/rasters"))
  output_files <- list.files("outputs/rasters")
  expect_equal(length(output_files), 2)
})

test_that("write_outputs Directory copies directory contents", {
  env <- setup_work_dir()
  on.exit(teardown_work_dir(env))

  src_dir <- tempfile("dir_")
  dir.create(src_dir)
  writeBin(charToRaw("file1"), file.path(src_dir, "a.txt"))
  writeBin(charToRaw("file2"), file.path(src_dir, "b.txt"))

  result <- Directory(path = src_dir)
  write_outputs(result)

  expect_true(dir.exists("outputs/directory"))
  output_files <- list.files("outputs/directory")
  expect_true("a.txt" %in% output_files)
  expect_true("b.txt" %in% output_files)
})

test_that("write_outputs preserves filename", {
  env <- setup_work_dir()
  on.exit(teardown_work_dir(env))

  src <- tempfile(tmpdir = tempdir(), pattern = "original_name_", fileext = ".tif")
  writeBin(charToRaw("data"), src)

  result <- File(path = src)
  write_outputs(result)

  output_files <- list.files("outputs/file")
  expect_equal(output_files, basename(src))
})

test_that("read_block_manifest returns NULL when no manifest", {
  env <- setup_work_dir()
  on.exit(teardown_work_dir(env))

  result <- read_block_manifest()
  expect_null(result)
})

test_that("read_block_manifest reads block.yaml", {
  env <- setup_work_dir()
  on.exit(teardown_work_dir(env))

  yaml::write_yaml(
    list(outputs = list(raster = list(type = "file"))),
    "block.yaml"
  )

  result <- read_block_manifest()
  expect_equal(result$raster$type, "file")
})

test_that("read_block_manifest reads from env var path", {
  env <- setup_work_dir()
  on.exit({
    Sys.unsetenv("SPADE_BLOCK_MANIFEST")
    teardown_work_dir(env)
  })

  manifest_file <- tempfile(fileext = ".yaml")
  yaml::write_yaml(
    list(outputs = list(custom = list(type = "file"))),
    manifest_file
  )
  Sys.setenv(SPADE_BLOCK_MANIFEST = manifest_file)

  result <- read_block_manifest()
  expect_equal(result$custom$type, "file")
})

test_that("infer_output_name for TabularFile", {
  expect_equal(infer_output_name(TabularFile(path = "/tmp/x")), "tabular")
})

test_that("infer_output_name for Directory", {
  expect_equal(infer_output_name(Directory(path = "/tmp/x")), "directory")
})

test_that("infer_output_name for VectorFileCollection", {
  expect_equal(infer_output_name(VectorFileCollection(paths = "/tmp/x")), "vectors")
})

test_that("infer_output_name for TabularFileCollection", {
  expect_equal(infer_output_name(TabularFileCollection(paths = "/tmp/x")), "tables")
})
